File: gateway/internal/handlers/creation_edit_attempts.go
package handlers

import (
	"encoding/json"
	"reflect"
	"strings"

	"github.com/aan/agent-assistant-gateway/internal/bridge"
	"github.com/aan/agent-assistant-gateway/internal/models"
)

// Read the immutable submitted operation, never infer it from today's plan.
// Legacy feedback is recoverable only while all candidate provenance exists.
func (h *CreationHandler) repairExecution(row models.CreationProject, nodeID string, candidates []string) *bridge.CreationRepairExecution {
	if len(candidates) == 0 {
		return nil
	}
	var result *bridge.CreationRepairExecution
	for _, id := range candidates {
		var run models.CreationRun
		if h.db.Table("creation_runs AS r").Select("r.*").Joins("JOIN creation_assets AS a ON a.run_id = r.id AND a.user_id = r.user_id").
			Where("a.id = ? AND a.user_id = ? AND r.project_id = ? AND r.project_node_id = ?", id, row.UserID, row.ID, nodeID).Take(&run).Error != nil {
			return nil
		}
		var graph creationGraph
		if json.Unmarshal([]byte(run.Definition), &graph) != nil || len(graph.Nodes) != 1 {
			return nil
		}
		node := graph.Nodes[0]
		if node.ID != nodeID || node.Kind != "image" {
			return nil
		}
		execution := &bridge.CreationRepairExecution{Operation: "generate"}
		switch node.ImageOperation {
		case "":
		case "edit":
			if len(node.AssetIDs) != 1 || node.AssetIDs[0] == "" {
				return nil
			}
			execution.Operation, execution.EditSourceAssetID = "edit", node.AssetIDs[0]
		default:
			return nil
		}
		if result != nil && !reflect.DeepEqual(result, execution) {
			return nil
		}
		result = execution
	}
	return result
}

func (h *CreationHandler) repairAttempts(row models.CreationProject, nodeID string, history []bridge.CreationRepairFeedback) []bridge.CreationRepairAttempt {
	result := []bridge.CreationRepairAttempt{}
	for _, earlier := range history {
		if earlier.NodeID != nodeID {
			continue
		}
		execution := earlier.Execution
		if execution == nil {
			execution = h.repairExecution(row, nodeID, earlier.CandidateIDs)
		}
		result = append(result, bridge.CreationRepairAttempt{Attempt: earlier.Attempt, CandidateIDs: earlier.CandidateIDs, Findings: earlier.Findings, Execution: execution})
	}
	if len(result) > 10 {
		result = result[len(result)-10:]
	}
	return result
}

// Defense in depth for a model response that ignores the operation budget.
// A different failed requirement or unknown legacy operation starts no chain.
func exhaustedEdits(req bridge.CreationPlanningRequest) bool {
	r := req.Repair
	if !req.AutomaticMode || r == nil || len(r.PreviousAttempts) < 2 {
		return false
	}
	for _, id := range req.LockedNodeIDs {
		if id == r.NodeID {
			return false
		}
	}
	node, exists := creativeNode(creativeDocument{Plan: req.CurrentPlan}, r.NodeID)
	if !exists || node.Kind != "image" {
		return false
	}
	chain := append([]bridge.CreationRepairAttempt{}, r.PreviousAttempts[len(r.PreviousAttempts)-2:]...)
	chain = append(chain, bridge.CreationRepairAttempt{Attempt: r.Attempt, CandidateIDs: r.CandidateIDs, Findings: r.Findings, Execution: r.Execution})
	type issue struct{ category, source, quote string }
	var common map[issue]bool
	for i, attempt := range chain {
		if attempt.Execution == nil || attempt.Execution.Operation != "edit" || attempt.Execution.EditSourceAssetID == "" || len(attempt.CandidateIDs) == 0 {
			return false
		}
		if i > 0 {
			linked := false
			for _, id := range chain[i-1].CandidateIDs {
				linked = linked || id == attempt.Execution.EditSourceAssetID
			}
			if !linked || attempt.Attempt != chain[i-1].Attempt+1 {
				return false
			}
		}
		candidates, reviewed := map[string]bool{}, map[string]bool{}
		for _, id := range attempt.CandidateIDs {
			candidates[id] = true
		}
		issues := map[issue]bool{}
		for _, finding := range attempt.Findings {
			reviewed[finding.CandidateID] = true
			issues[issue{finding.Category, finding.SourceID, finding.RequirementQuote}] = true
		}
		if !reflect.DeepEqual(candidates, reviewed) {
			return false
		}
		if common == nil {
			common = issues
		} else {
			next := map[issue]bool{}
			for old := range common {
				for current := range issues {
					if old.category == current.category && old.source == current.source && old.quote != "" && current.quote != "" &&
						(strings.Contains(old.quote, current.quote) || strings.Contains(current.quote, old.quote)) {
						if len(old.quote) < len(current.quote) {
							next[old] = true
						} else {
							next[current] = true
						}
					}
				}
			}
			common = next
		}
	}
	return len(common) > 0
}

File: agent/aigc/creation_edit_attempts.py
"""Bound unsuccessful edit chains using actual execution and review history."""


def exhausted_edit_task(request):
    repair = request.repair
    if not request.automatic_mode or not repair or len(repair.previous_attempts) < 2:
        return None
    node = next((n for n in request.current_plan.get('nodes', []) if n['id'] == repair.node_id), None)
    if not node or node.get('kind') != 'image' or node['id'] in request.locked_node_ids:
        return None
    chain = [*repair.previous_attempts[-2:], repair]
    persisting = None
    for index, attempt in enumerate(chain):
        execution = attempt.execution
        candidates = set(attempt.candidate_ids)
        if (not execution or execution.operation != 'edit' or not execution.edit_source_asset_id
                or not candidates or {f.candidate_id for f in attempt.findings} != candidates):
            return None
        if index and (attempt.attempt != chain[index - 1].attempt + 1
                or execution.edit_source_asset_id not in chain[index - 1].candidate_ids):
            return None
        # A category alone can describe unrelated requirements. A reviewer may
        # quote a sentence or its exact subclause; match nested literal quotes.
        issues = {(f.category, f.source_id, f.requirement_quote) for f in attempt.findings}
        persisting = issues if persisting is None else {
            (old[0], old[1], min((old[2], new[2]), key=len))
            for old in persisting for new in issues
            if old[:2] == new[:2] and old[2] and new[2]
            and (old[2] in new[2] or new[2] in old[2])}
    if not persisting:
        return None
    return dict(operation='replan_generation',
        persisting_categories=sorted({issue[0] for issue in persisting}),
        instruction='连续三次编辑仍未通过同一验收要求。本轮必须更换操作：清空该节点edit_source_asset_id，'
        '基于原始验收要求、已确认参考和实际失败历史重新设计生成方案，必要时增加明确分工的前置参考。'
        '不能再把上一张失败图作为编辑来源重复提交，不能只把原编辑指令换一种说法。'
        '保留冻结content、锁定节点及交付数量，不放宽验收、不把失败图提升为参考。新候选仍需完整复审。')
